server/src/main/java/org/elasticsearch/search/profile/query/MultiProfileCollectorManager.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.profile.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * A {@link CollectorManager} that wraps instances created by another CollectorManager before passing them on.
+ * It delegates all the profiling to the generated collectors via {@link #getCollectorTree()} and joins them up in one
+ * {@link CollectorResult}.
+ */
+public final class MultiProfileCollectorManager<T> implements CollectorManager<Collector, T> {
+
+    protected final CollectorManager<Collector, T> collectorManager;
+    private final List<Collector> wrappedCollectors = new ArrayList<>();
+    private final List<InternalProfileCollector> profileCollectors = new ArrayList<>();
+
+    public MultiProfileCollectorManager(CollectorManager<Collector, T> collectorManager) {
+        this.collectorManager = collectorManager;
+    }
+
+    @Override
+    public Collector newCollector() throws IOException {
+        Collector collector = collectorManager.newCollector();
+        wrappedCollectors.add(collector);
+        InternalProfileCollector internalProfileCollector = new InternalProfileCollector(collector, CollectorResult.REASON_SEARCH_TOP_HITS);
+        profileCollectors.add(internalProfileCollector);
+        return internalProfileCollector;
+    }
+
+    public T reduce(Collection<Collector> collectors) throws IOException {
+        return collectorManager.reduce(wrappedCollectors);
```

Comment:
this is somehow controversial: I think you should make this CollectorManager typed on InternalProfileCollector, because that's what it returns. Then unwrap in reduce (although keeping track of inner collectors should achieve the same).

I am also thinking that the logic in getCollectorTree should be done on reduce, which sets an instance member which getCollectorTree returns. That way you don't need to keep a reference to profile collectors and wrapped collectors.